File: test_SfSNet.m
clc; clear;
%% TO DO: Add your Matcaffe path as $PATH_TO_CAFFE/matlab
addpath(genpath('/scratch2/photometric_stereo/caffe/matlab/'));

addpath(genpath('functions'));

model = 'SfSNet_deploy.prototxt';
weights='SfSNet.caffemodel.h5';

GPU_ID=0; %Set your GPU ID
caffe.set_mode_gpu();
caffe.set_device(GPU_ID);
net = caffe.Net(model, weights, 'test');


%% Choose Dataset
%Images and masks are provided
%list_im=dir('Images_mask/*_face.png'); dat_idx=1;


%No mask provided (Need to use your own mask).
list_im=dir('Images/*.png'); dat_idx=0; %Uncomment to test with this mode

M=128; %size of input for SfSNet
for i=1:length(list_im)
    
    if dat_idx
        im=imread(['Images_mask/' list_im(i).name]); im=imresize(im,[M M]);
        mask_name=['Images_mask/' strrep(list_im(i).name,'face','mask')];
        Mask=imread(mask_name); mask=single(Mask)/255; mask=imresize(mask,[M M]);
    else
        im=imread(['Images/' list_im(i).name]); im=imresize(im,[M M]);
    end
    
    %Prepare images
   im=reshape(im,[size(im)]); im=single(im)/255;
    im_data = im(:, :, [3, 2, 1]);  % permute channels from RGB to BGR
    im_data = permute(im_data, [2, 1, 3]); 
    
    %pass images  
    out_im = net.forward({im_data});
    n_out=out_im{2}; al_out=out_im{1}; light_out=out_im{3};
    
    %light_out is a 27 dimensional vector. 9 dimension for each channel of
    %RGB. For every 9 dimensional, 1st dimension is ambient illumination
    %(0th order), next 3 dimension is directional (1st order), next 5
    %dimension is 2nd order approximation. You can simply use 27
    %dimensional feature vector as lighting representation.
    
    %Transform
    n_out2=n_out(:,:,[3 2 1]); 
    n_out2=imrotate(n_out2,-90); n_out2=fliplr(n_out2);
    n_out2=2*n_out2-1; %[-1 1]
    nr=sqrt(sum(n_out2.^2,3)); n_out2=n_out2./repmat(nr,[1 1 3]);
    
    al_out2=imrotate(al_out,-90);
    al_out2=al_out2(:,:,[3 2 1]); al_out2=fliplr(al_out2);
    
    %% Note: n_out2, al_out2, light_out is the actual output
    
    [Irec,Ishd]=create_shading_recon(n_out2,al_out2,light_out);
    
    
    if dat_idx
        subplot(2,3,1); imshow(im); title('Image');
        subplot(2,3,2); imshow(((1+n_out2)/2).*mask + (1-mask).*ones(M,M,3)); title('Normal');
        subplot(2,3,3); imshow(al_out2.*mask+ (1-mask).*ones(M,M,3)); title('Albedo');
        subplot(2,3,5); imshow(Ishd.*mask+ (1-mask).*ones(M,M,3)); title('Shading');
        subplot(2,3,6); imshow(Irec.*mask+ (1-mask).*ones(M,M,3)); title('Recon');
    else
        subplot(2,3,1); imshow(im); title('Image');
        subplot(2,3,2); imshow((1+n_out2)/2); title('Normal');
        subplot(2,3,3); imshow(al_out2);  title('Albedo');
        subplot(2,3,5); imshow(Ishd); title('Shading');
        subplot(2,3,6); imshow(Irec); title('Recon');
    end
      
        disp('Press Enter to Continue');
        pause();
end

    
    
    
    

File: test_SfSNet.py
import os
import sys
import numpy as np
import matplotlib.pyplot as plt
plt.ion()
import cv2
import caffe
import spherical_harmonics

# caffe.set_mode_cpu();
caffe.set_device(0);
caffe.set_mode_gpu();

model = 'SfSNet_deploy.prototxt'
weights = 'SfSNet.caffemodel.h5'
net = caffe.Net(model, weights, caffe.TEST)

# Choose Dataset
dat_idx = input('Please enter 1 for images with masks and 0 for images without mask: ')
dat_idx = int(dat_idx)
if dat_idx == 1:
    # Images and masks are provided
    list_im = os.listdir('Images_mask/');
    list_im = list(filter(lambda f: f.endswith('_face.png'), list_im))
    list_im.sort()
elif dat_idx == 0:
    # No mask provided (Need to use your own mask).
    list_im = os.listdir('Images/');
    list_im = list(filter(lambda f: f.endswith('.png'), list_im))
    list_im.sort()
else:
    print('Wrong Option!');
    sys.exit(1)

fig, axs = plt.subplots(2, 3, figsize=(24,16))
for i in range(2):
    for j in range(3):
        axs[i][j].axis('off')
axs[0][0].set_title('Image')
axs[0][1].set_title('Normal')
axs[0][2].set_title('Albedo')
axs[1][1].set_title('Shading')
axs[1][2].set_title('Recon')

M = 128; # size of input for SfSNet
for i in range(len(list_im)):
    if dat_idx == 1:
        im = cv2.imread(os.path.join('Images_mask/', list_im[i]));
        im = cv2.cvtColor(im, cv2.COLOR_BGR2RGB)

        mask = cv2.imread(os.path.join('Images_mask/', list_im[i].replace('face', 'mask')));
        mask = cv2.resize(mask, (M, M));
        mask = mask/255;
    else:
        im = cv2.imread(os.path.join('Images/', list_im[i]));
        im = cv2.cvtColor(im, cv2.COLOR_BGR2RGB)

    # Prepare images
    im = cv2.resize(im, (M, M))
    im = im.astype(np.float32)/255
    im_data = cv2.cvtColor(im, cv2.COLOR_RGB2BGR)
    im_data = im_data.reshape((1,) + im_data.shape)
    im_data = im_data.transpose((0,3,1,2))

    # Pass images
    net.blobs['data'].data[...] = im_data
    output = net.forward();
    al_out = output['Aconv0'].transpose((0,2,3,1))[0]
    n_out  = output['Nconv0'].transpose((0,2,3,1))[0]
    light_out = output['fc_light'][0]

    # light_out is a 27 dimensional vector. 9 dimension for each channel of
    # RGB. For every 9 dimensional, 1st dimension is ambient illumination
    # (0th order), next 3 dimension is directional (1st order), next 5
    # dimension is 2nd order approximation. You can simply use 27
    # dimensional feature vector as lighting representation.

    # Transform
    n_out2 = cv2.cvtColor(n_out, cv2.COLOR_BGR2RGB)
    n_out2 = 2*n_out2-1; # [-1 1]
    n_out2 = n_out2/(np.expand_dims(np.linalg.norm(n_out2,axis=2),axis=2)+1e-10)

    al_out2 = cv2.cvtColor(al_out, cv2.COLOR_BGR2RGB)

    # Note: n_out2, al_out2, light_out is the actual output

    # Create reconstruction and shading image
    light_out_data = light_out.reshape((1,) + light_out.shape)
    n_out2_data = n_out2.reshape((1,) + n_out2.shape)
    Ishd = spherical_harmonics.compute_shading(light_out_data, n_out2_data)[0]
    Irec = Ishd*al_out2

    # Visualize light_out on a sphere (not included in test_SfSNet.m)
    sphere_normals = spherical_harmonics.get_sphere_normals()
    sphere_normals_data = sphere_normals.reshape((1,) + sphere_normals.shape)
    Ilight = spherical_harmonics.compute_shading(light_out_data, sphere_normals_data)[0]

    if dat_idx == 1:
        axs[0][0].imshow(mask*im)
        axs[0][1].imshow(mask*((1+n_out2)/2).clip(0,1))
        axs[0][2].imshow(mask*al_out2.clip(0,1))
        axs[1][0].imshow(Ilight.clip(0,1))
        axs[1][1].imshow(mask*200/255*Ishd.clip(0,1))
        axs[1][2].imshow(mask*Irec.clip(0,1))
    else:
        axs[0][0].imshow(im)
        axs[0][1].imshow(((1+n_out2)/2).clip(0,1))
        axs[0][2].imshow(al_out2.clip(0,1))
        axs[1][0].imshow(Ilight.clip(0,1))
        axs[1][1].imshow(200/255*Ishd.clip(0,1))
        axs[1][2].imshow(Irec.clip(0,1))

    input('Press Enter to Continue')
